Added fix for case where CCP swagger returns result with outdate Expires: header

cbz opened the pull request:
Sometimes the EVE API returns results with an Expiry date set to the past - currently this causes esipy to exception, as the negative value calculated for cache_timeout is passed onto redis which throws an exception.

I'm just adding a check for this case and a few warnings so this can be diagnosed if it occurs.

Kyria commented:
Although I don't really like that fix, because that case is not supposed to happen, I understand the need (since it breaks everything).

Can you add test to test invalid expires and fix your code please (as travis tests are yelling) ? 
Then i'll merge it. 

```
$ flake8 esipy
esipy/client.py:272:80: E501 line too long (91 > 79 characters)
esipy/client.py:285:80: E501 line too long (88 > 79 characters)
esipy/client.py:287:1: W391 blank line at end of file
``` 

Thanks !

cbz commented:
So I have fixed all of these issues and may issue another pull request in a bit, that said testing with Travis-CI has uncovered a strange bug that appears to be latent.  I added the following test:

`    def test_esipy_expired_response(self):
        operation = self.app.op['get_incursions']

        with httmock.HTTMock(public_incursion_expired):
            warnings.simplefilter('ignore')
            incursions = self.client_no_auth.request(operation())
            self.assertEqual(incursions.status, 200)

            warnings.resetwarnings()
            warnings.simplefilter('error')

            with self.assertRaises(UserWarning):
                incursions = self.client_no_auth.request(operation())
`
and this passes testing, except on versions of python before 3.4 - the error is that the second assert fails with "AssertionError: UserWarning not raised". After a lot of testing simply commenting out the first call of self.client_no_auth.... has this test case passing.  At this point I'm pretty sure it's not my additions. 

If I edit an existing test case like so:

`    def test_client_warning_header(self):
        with httmock.HTTMock(public_incursion_warning):
            incursion_operation = self.app.op['get_incursions']
            self.client_no_auth.request(incursion_operation())
            warnings.simplefilter('error')

            with self.assertRaises(UserWarning):
                self.client_no_auth.request(incursion_operation())
`
This will fail with an identical error but only under pythons <= 3.3 . If I remove the first call to client_no_auth it will again pass.

Kyria commented:
That's... interesting. I'm now wondering is assertRaises is the best way to check warnings... found this https://docs.python.org/2.7/library/warnings.html#testing-warnings 

This could be a warning behavior issue in py2.7/3.4 that, looking at your code above, you raise once / ignore once raised warning, and then it doesn't throw it again.
It'd probably be fixed if you try first the test with raise, and after the test "i don't want to see anything" (which is the warning behavior after all: raise only once, unless you specify otherwise) : 

```python
def test_esipy_expired_response(self):
        operation = self.app.op['get_incursions']

        with httmock.HTTMock(public_incursion_expired):
            warnings.simplefilter('error')

            with self.assertRaises(UserWarning):
                incursions = self.client_no_auth.request(operation())

            warnings.resetwarnings()
            warnings.simplefilter('ignore')
            incursions = self.client_no_auth.request(operation())
            self.assertEqual(incursions.status, 200)
```
Didn't test it right now but you get the idea. 
I'll give a review what you've all done and merge everything tomorrow :) 

Thanks !

Kyria commented:
fyi, i released your fix within the 0.3.1 release.
